### DISCUSSION TITLE
Return Data Object based on row count

### DISCUSSION THREAD

**Discussion by sharoonamjid** [simolus3/drift#1282] [Q&A] [answered]
I have a table called: `DownloadMediaData`

```
@DataClassName("DownloadMediaData")
class DownloadsMediaData extends Table {
  IntColumn get mediaId => integer()();

  TextColumn get mediaType => text()();

  TextColumn get mediaFileUrl => text()();

  TextColumn get mediaName => text()();

  IntColumn get mediaDownloadStatus => integer()
      .withDefault(Variable.withInt(MediaDownloadStatus.enqueued.value))
      .nullable()();

  DateTimeColumn get createdAt => dateTime()
      .withDefault(Variable.withDateTime(DateTime.now()))
      .nullable()();

  @override
  Set<Column> get primaryKey => {mediaFileUrl};
}
```

The int column `mediaDownloadStatus` defined in above table represent the current status of download, it can be `1` which represents `enqueued` status or it can be `2` which represents `downloading` status

What I want is to be able to get maximum three `enqueued` `DownloadMediaData` only if `downloading` rows are less than 4. So max 3 files will be sent to download and countless can be enqueued.

**Comment by simolus3** [comment, marked as answer]
Do variables in `withDefault` actually work? I think it's better to use a `Constant<int>` there instead.

Regarding your query, couldn't you just `orderBy([OrderingTerm.desc(downloadsMediaData.mediaDownloadStatus)])` and then `limit(3)`? Then you'd get the media that's `downloading` first, followed by some media that is enqueued. You could filter out the things that are already downloading in Dart later.

There's a way to do this in SQL by using a subquery in a limit clause, but that's not currently available in moor's Dart api. I can take a look at adding it though.

**Reply by sharoonamjid** [reply]
> Do variables in withDefault actually work? I think it's better to use a Constant<int> there instead.

I tried using `Constant` and it was working for `int` but I didn't know how to use `DateTime.now()` inside `withDefault` for createdAt `DateTimeColumn` column.

> Regarding your query, couldn't you just orderBy([OrderingTerm.desc(downloadsMediaData.mediaDownloadStatus)]) and then limit(3)

I didn't get your point. Can you please help me with a snipped please? I want max 3 `enqueued` media no matter how many gets are called from this table. In my dart code once the `enqueued` media are received from the table then it is sent for download and the status is updated to `downloading` from `enqueued`. if `downloading` rows are reached to 3 in table, then I don't want to return any `enqueued` row because I want to allow only 3 files to be downloaded simultaneously.

**Reply by sharoonamjid** [reply]
I got your approach of `orderyBy` `downloading` with `limit(3)`. Thank you

**Reply by simolus3** [reply]
Note that the `DateTime.now()` will be called when the database is opened, not when you insert an entry. You could either use `clientDefault(() => DateTime.now())` or `withDefault(currentDateAndTime)`.

**Reply by sharoonamjid** [reply]
Awesome. Thanks.
